src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -248,22 +257,8 @@ public void Reset()
 
         protected void ResetComponents()
         {
-            var frameHeaders = Interlocked.Exchange(ref _frameHeaders, null);
-            if (frameHeaders != null)
-            {
-                RequestHeaders = null;
-                ResponseHeaders = null;
-                HttpComponentFactory.DisposeHeaders(frameHeaders);
-            }
-
-            var frameStreams = Interlocked.Exchange(ref _frameStreams, null);
-            if (frameStreams != null)
-            {
-                RequestBody = null;
-                ResponseBody = null;
-                DuplexStream = null;
-                HttpComponentFactory.DisposeStreams(frameStreams);
-            }
+            _frameHeaders?.Uninitialize();
+            _frameStreams?.Uninitialize();
```

Comment:
Uninitialized does nothing for streams since StopStreams should have already been called.